### PR TITLE
Simplify Hashable implementation

### PR DESCRIPTION
**Pull request by Ehlster**
With the current implementation, nodes with the same type and contents have the same hash. But then in the `Equatable` implementation, we check for the same `cmark_node`. Therefore only `Node` instances for the same `cmark_node` can be used as the key in a dictionary or set, making the `Hashable` implementation more complicated than it needs to be.

With this change we also use the `cmark_node` for the `Hashable` implementation.

### CODE REVIEW

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (mattt) [inline Tests/CommonMarkTests/NodeHashableTests.swift]: Is there any reason not to assign the same text into a common variable? If the concern has to do with initialization from a single value, we could achieve the same effect with string interpolation:

```suggestion
        let text = #"""
        All human beings are born free and equal in dignity and rights.
        They are endowed with reason and conscience
        and should act towards one another in a spirit of brotherhood.
        """#
        
        let firstNode = Paragraph(text: "\(text)", replacingNewLinesWithBreaks: true)
        let secondNode = Paragraph(text: "\(text)", replacingNewLinesWithBreaks: true)
```

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (mattt) [inline Tests/CommonMarkTests/NodeHashableTests.swift]: I bumped on the name `NodeHashableTests`, despite `Equatable` indeed being a requirement of `Hashable`. I think I'd prefer `NodeTests`, but I'd probably also want to consolidate the various `Document___Tests` into a single file as well. We can do that in a follow-up; no need to change that here. Any strong preference one way or another?

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (mattt) [inline Tests/CommonMarkTests/NodeHashableTests.swift]: Although the meaning is more precise within the context of the test case, I like for test case names to be more self-descriptive. I find XCTest's default test output is hard to parse (both for humans and computers), so I tend to design test names to make it easy to tell what failed.

Here's one suggestion: 

```suggestion
    func testHashValueConsistentAfterModification() {
```

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (mattt) [inline Tests/CommonMarkTests/NodeHashableTests.swift]: I think the intent of the test would be clearer if we used `Set<Node>` instead of `[Node: String]`.

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (mattt) [inline Tests/CommonMarkTests/NodeHashableTests.swift]: Just for completeness, we should add an assertion on the final text of the paragraph.

**Review by mattt — CHANGES_REQUESTED** [review]
Thanks for putting this PR together, @Lukas-Stuehrk. Great work! You're indeed correct that our current implementation of `Hashable` does a lot of unnecessary work — and worse, produces incorrect results. I have to think that this was a holdover from an earlier sketch where I was going for value semantics instead of reference semantics.

Just a few requested changes, all trivial matters about the tests themselves. The implementation itself is perfect.

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (Ehlster) [reply Tests/CommonMarkTests/NodeHashableTests.swift]: I had similar thoughts, but I wanted to keep the name of the file consistent with the rest of the tests. +1 for merging into a single file.

Also, I added the tests for the `Hashable` implemtation only for `Node`. We should add them for `Document` as well, if we want to have full test coverage for the publicly exposed API of the library. The fact that `Document` inherits from `Node` and does not override the `Hashable` implementation is an implementation detail which could change in the future. Do you have any strong opinion on this?

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (Ehlster) [reply Tests/CommonMarkTests/NodeHashableTests.swift]: Hahaha, this feels as if you could see the troubles I had in finding a good name for this test case. 👍 That's a perfect name for it, I will take it.

File: Tests/CommonMarkTests/NodeHashableTests.swift
Comment (Ehlster) [reply Tests/CommonMarkTests/NodeHashableTests.swift]: I renamed the file (and class name) to `NodeTests` and organized the tests for the different parts with `// MARK: -` Then we can add more tests for `Node` in the future and unify the tests for `Document` with a later change.

Comment (mattt) [reply Tests/CommonMarkTests/NodeHashableTests.swift]: > Also, I added the tests for the `Hashable` implemtation only for `Node`. We should add them for `Document` as well, if we want to have full test coverage for the publicly exposed API of the library. The fact that `Document` inherits from `Node` and does not override the `Hashable` implementation is an implementation detail which could change in the future. Do you have any strong opinion on this?

Since `Document` was originally _not_ a subclass of `Node`, I think it's reasonable to have test coverage in case that ever changes again. I'll do that in a follow-up after this gets merged.

> I renamed the file (and class name) to NodeTests and organized the tests for the different parts with // MARK: - Then we can add more tests for Node in the future and unify the tests for Document with a later change.

Perfect. Thanks!

**Review by mattt — APPROVED** [review]

**Comment by mattt** [comment]
Alright, I added a changelog entry with 78c19c0, so this will be good to merge once the latest CI run passes. Thanks again for your help with this, @Lukas-Stuehrk!

**Pull request merged**
